tools/apps/src/bin/quiche-server.rs

```diff
@@ -0,0 +1,504 @@
+                        )
+                        .unwrap();
+                        let out = &out[..len];
+
+                        if let Err(e) = socket.send_to(out, &src) {
+                            if e.kind() == std::io::ErrorKind::WouldBlock {
+                                debug!("send() would block");
+                                break;
+                            }
+
+                            panic!("send() failed: {:?}", e);
+                        }
+                        continue;
+                    }
+
+                    odcid = validate_token(&src, token);
+
+                    // The token was not valid, meaning the retry failed, so
+                    // drop the packet.
+                    if odcid == None {
+                        error!("Invalid address validation token");
+                        continue;
+                    }
+
+                    if scid.len() != hdr.dcid.len() {
+                        error!("Invalid destination connection ID");
+                        continue;
+                    }
+
+                    // Reuse the source connection ID we sent in the Retry
+                    // packet, instead of changing it again.
+                    scid.copy_from_slice(&hdr.dcid);
+                }
+
+                debug!(
+                    "New connection: dcid={} scid={}",
+                    hex_dump(&hdr.dcid),
+                    hex_dump(&scid)
+                );
+
+                let conn = quiche::accept(&scid, odcid, &mut config).unwrap();
+
+                let client = Client {
+                    conn,
+                    http_conn: None,
+                    partial_responses: HashMap::new(),
+                };
+
+                clients.insert(scid.to_vec(), (src, client));
+
+                clients.get_mut(&scid[..]).unwrap()
+            } else {
+                clients.get_mut(&hdr.dcid).unwrap()
+            };
+
+            // Process potentially coalesced packets.
+            let read = match client.conn.recv(pkt_buf) {
+                Ok(v) => v,
+
+                Err(quiche::Error::Done) => {
+                    debug!("{} done reading", client.conn.trace_id());
+                    break;
+                },
+
+                Err(e) => {
+                    error!("{} recv failed: {:?}", client.conn.trace_id(), e);
+                    break 'read;
+                },
+            };
+
+            debug!("{} processed {} bytes", client.conn.trace_id(), read);
+
+            // Create a new HTTP connection as soon as the QUIC connection
+            // is established.
+            if client.http_conn.is_none() &&
+                (client.conn.is_in_early_data() ||
+                    client.conn.is_established())
+            {
+                let app_proto = client.conn.application_proto();
+                let app_proto = &std::str::from_utf8(&app_proto).unwrap();
```

Comment:
Since I need value, do you want me to pattern match rather than unwrap?

Reply:
Oh, wait, I just realized that this is always going to be valid, otherwise we close the connection because of failed ALPN negotiation, so `unwrap()` is probably fine here, though maybe add a comment.